hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/processor/HashJoinP.java

```diff
@@ -107,9 +104,38 @@ protected boolean tryProcess0(@Nonnull Object item) {
         return result == null || tryEmit(result);
     }
 
-    @Nullable
-    private Object lookupJoined(int ordinal, E0 item) {
-        return lookupTables.get(ordinal).get(keyFns.get(ordinal).apply(item));
+    private boolean tryProcessWhenNoTags(E0 e0) {
+        boolean allItemsAccepted = true;
+        boolean noItemsAtAll = true;
+        if (keyFns.size() == 2) {
+            for (Object joinedObject : lookupJoined(1, e0)) {
+                noItemsAtAll = false;
+                Object remapped = mapToOutputBiFn.apply(e0, joinedObject);
+
+                allItemsAccepted &= tryEmit(remapped);
+            }
+        } else {
+            for (Object firstJoinedObject : lookupJoined(1, e0)) {
+                for (Object secondJoinedObject : lookupJoined(2, e0)) {
+                    noItemsAtAll = false;
+                    Object remapped = mapToOutputTriFn.apply(e0, firstJoinedObject, secondJoinedObject);
+
+                    allItemsAccepted &= tryEmit(remapped);
+                }
+            }
+        }
+
+        return noItemsAtAll || allItemsAccepted;
```

Comment:
If part of the items is accepted, when `tryProcess` is called next time you should retry offering the same item. It might be easier if you use a `Traverser`. You should also write a unit test using `TestSupport` which tests the cooperative emission and simulates a full outbox after every item.

Reply:
Ok, will try to rewrite it to Traversers

Reply:
Also have a look at `AbstractProcessor.flatMapper`: You'll basically create a function that will create a traverser for each input item and then simply call `return flatMapper.tryProcess(item)` for each item. Have a look at [`TransfomP`](https://github.com/hazelcast/hazelcast-jet/blob/fe66afd5f51331aee626aa273a622cdfef3e1fb6/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/processor/TransformP.java#L32-L46) (the processor that implements `flatMap`).

This might be the easiest way.

Reply:
@viliam-durina Thanks for your comment, I've sent a change to use flatMapper and some tests